File: python3/loops.py
from random import random
import time

def simulate():
    start = time.time()
    iterations = 100_000
    sample_size = 23

    count = 0
    for x in range(iterations):
        data = []

        for i in range(sample_size):
            rand = int(random() * 365)
            if rand in data:
                count += 1;
                break
            else:
                data.append(rand)

    print("iterations:", iterations)
    print("sample-size:", sample_size)
    results = round(count / iterations * 100, 2);
    print("percent:", results)
    end = time.time()
    diff = round(end-start, 3)
    print("seconds:", diff)

simulate()

File: haskell/loops.hs
import System.Random
import Data.Time.Clock.POSIX
import Data.List.Split
import qualified Data.IntSet as IntSet
import System.Environment

getTimeMillis :: (Integral a) => IO a
getTimeMillis = round <$> (*1000) <$> getPOSIXTime

roundTo n =
  let shifter = 10^n
  in (/shifter) . fromIntegral . round . (*shifter)

main = do
  start <- getTimeMillis
  iterations <- read <$> head <$> getArgs :: IO Int
  gen <- getStdGen
  let sampleSize = 23 :: Int
      range = (0, 364) :: (Int, Int)
      results = (/ (fromIntegral iterations)) . fromIntegral . simulate iterations sampleSize range $ gen
  putStrLn . ("iterations: " ++) . show $ iterations
  putStrLn . ("sample-size: " ++) . show $ 23
  putStrLn . ("percent: " ++) . show . roundTo 2 . (*100) $ results
  end <- getTimeMillis
  putStrLn . ("seconds: " ++) . show . (/1000) . fromIntegral $ (end - start)

simulate :: Int -> Int -> (Int, Int) -> StdGen -> Int
simulate iterations sampleSize range =
  let generateLists = take iterations . chunksOf sampleSize . randomRs range
      listHasDuplicates = (/= sampleSize) . IntSet.size . IntSet.fromList
  in length . filter listHasDuplicates . generateLists

-- iterations       100000
-- sample-size      23
----------------------------
-- data-structure   time (s)
----------------------------
-- Infinite List    0.918
-- Full IntSet      1.040
-- Partial IntSet   1.069
-- Partial List     1.142
-- Full List        1.209
-- Partial Array    4.470
